File: services/meta-assembly/app/utils/logger.py
import json
import time
import uuid
from typing import Any

from app.config import settings


class StructuredLogger:
    """Structured JSON logger — no bare print() statements allowed."""

    def __init__(self, layer: str = "service"):
        self.service = settings.service_name
        self.layer = layer

    def _emit(self, level: str, message: str, **extra: Any) -> None:
        record: dict[str, Any] = {
            "level": level,
            "message": message,
            "service": self.service,
            "layer": self.layer,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "span_id": str(uuid.uuid4()),
        }
        record.update(extra)
        print(json.dumps(record))  # stdout → collected by Docker/k8s

    def debug(self, message: str, **extra: Any) -> None:
        self._emit("DEBUG", message, **extra)

    def info(self, message: str, **extra: Any) -> None:
        self._emit("INFO", message, **extra)

    def warning(self, message: str, **extra: Any) -> None:
        self._emit("WARNING", message, **extra)

    def error(self, message: str, **extra: Any) -> None:
        self._emit("ERROR", message, **extra)

    def critical(self, message: str, **extra: Any) -> None:
        self._emit("CRITICAL", message, **extra)


def get_logger(layer: str = "service") -> StructuredLogger:
    return StructuredLogger(layer=layer)

File: services/agent-mgmt/app/utils/logger.ts
const SERVICE_NAME = 'agent-mgmt';

type LogLevel = 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';

interface LogRecord {
  level: LogLevel;
  message: string;
  service: string;
  layer: string;
  timestamp: string;
  trace_id?: string;
  span_id?: string;
  duration_ms?: number;
  [key: string]: unknown;
}

function log(level: LogLevel, message: string, layer: string, extra: Record<string, unknown> = {}): void {
  const record: LogRecord = {
    level,
    message,
    service: SERVICE_NAME,
    layer,
    timestamp: new Date().toISOString(),
    ...extra,
  };
  // stdout only — collected by Docker/k8s log aggregator
  process.stdout.write(JSON.stringify(record) + '\n');
}

export const logger = {
  debug: (message: string, layer: string, extra?: Record<string, unknown>) =>
    log('DEBUG', message, layer, extra),
  info: (message: string, layer: string, extra?: Record<string, unknown>) =>
    log('INFO', message, layer, extra),
  warn: (message: string, layer: string, extra?: Record<string, unknown>) =>
    log('WARN', message, layer, extra),
  error: (message: string, layer: string, extra?: Record<string, unknown>) =>
    log('ERROR', message, layer, extra),
};
